### ISSUE TITLE
orderRef not used?

### ISSUE THREAD

**Issue by mike4242** [pilwon/node-ib#110]
Hi,

I set the property orderRef on the order object, before I call placeOrder. However, the orderRef doesn't seem to have been included, when I pull out history as a csv file from IB. 

The call goes like this: 

    ib.order.orderRef = "Testing orderref abc123";
    ib.placeOrder(orderid, ib.contract.stock(ticker), ib.order.limit('BUY', buyamount, the_price));

Am I supposed to supply the orderRef in a different way, rather than setting the property as above? 

Regards,
Mike

**Comment by nemozny** [comment]
OrderRef works, only you created two different ib.order objects and sent the second one.
The correct sequence is:
```javascript
let LMT = ib.order.limit('BUY', buyamount, the_price);
LMT.orderRef = "Testing orderref abc123";
ib.placeOrder(orderid, contract, LMT);
```
Pilwons ib.order.limit actually just creates order object with orderType = LMT. 
See https://github.com/pilwon/node-ib/blob/master/lib/order/limit.js

**Comment by nemozny** [comment]
You can close this.

**Comment by mike4242** [comment]
Yes, you are absolutely right. I discovered that, but forgot to post here. BUT I tried the correct one as well previously and neither that worked and the reason is that OrderRef is not supported on virtual paper trading accounts. You won't be able to retrieve it afterwards, when you log in and create a document that contains all transations. However, for real accounts, you will.

**Comment by nemozny** [comment]
Hi @mike4242 , no, it works, I am using paper account and this is what I just submitted to IB via API.

![orderref](https://user-images.githubusercontent.com/1666780/33436882-92b61466-d5e6-11e7-8c2c-ca07e8c266c8.png)

**Comment by nemozny** [comment]
The real issue here (for me) is that it is really just an ORDER, not POSITION tag. So you can recover orderRef only using reqExecutions and executions are available for only 24 hours or so.
Anyway you can create bracket order with MKT and STP orders and tag both. After STP gets automatically executed by IB, you can call reqExecutions and there you can see STP orderRef.
The thing is you have to know, that your STP order got executed. You have 2 options:
1. run reqExecutions at least every 24 hours (long time in between)
2. forget executions and track and count your positions via ib.reqAutoOpenOrders(true), which is a subscription, so you will know changes in positions immediately.

I chose the latter.
Cheers

**Comment by nemozny** [comment]
YES! Even ALGOS do work! Or at least adaptive.
```javascript
let order = {
		action : "BUY",
		orderType : "MKT",
		algoStrategy : 'Adaptive',
		algoParams : {"adaptivePriority" : "Normal" },
		transmit : true
	}
```
Documentation - https://interactivebrokers.github.io/tws-api/ibalgos.html#gsc.tab=0

**Comment by mike4242** [comment]
@nemozny Keep in mind that reqAutoOpenOrders applies to the account rather than a specific client ID. So if you use more than one client and they trade the same security, then it won't work. 

You can check the fill parameter in the orderStatus callback to see if it got executed (filled, and if so, how many).

**Comment by mike4242** [comment]
@nemozny With Adaptive Algo, don't you have to include a base order as well? Which includes, for example, the number of stocks (or whatever) you want to sell? 

base_order =  ib.order.limit('SELL', numToSell, thePrice);
base_order.algoStrategy = 'Adaptive';
base_order.algoParams = {"adaptivePriority" : "Normal" },
etc.
???

**Comment by nemozny** [comment]
@mike4242 Yes.

**Comment by amcintyre99** [comment]
I can't get this line to work as written above:
algoParams : {"adaptivePriority" : "Normal" },

After testing I figured out this line in outgoing.js is checking for an array and since it's not, it gets 0:
algoParamsCount = (!_.isArray(order.algoParams) ? 0 : order.algoParams.length);

I've tried turning the algoParams line into an array but I'm not much of a JS person so have failed miserably.

Any advice? thanks!!

**Comment by nemozny** [comment]
@amcintyre99 Yeah, sorry, I didn't try to change the priority. It seems the "Normal" is the default setting here, and when you type it incorrectly as in my example above.

The correct syntax is:
let order = {
		action : "BUY",
		orderType : "MKT",
		algoStrategy : 'Adaptive',
		algoParams : [ {tag : "adaptivePriority", value : "Normal" } ],
		transmit : true
};

Tested and working.
BR

**Comment by mike4242** [comment]
Can you get good prices on stocks? For MKT prices, one would expect to get only one or two pips worse than the current bid or ask. However, the results I got were a lot worse. What are your experiences? For currencies, I always get perfect prices. Often equal to bid when selling or ask when buying. Which is the way it should be. However, for stocks I get so much worse prices. Lots of slippage... Any tips on how to improve that bit? 

/Mike

**Comment by amcintyre99** [comment]
Thanks @nemozny , I had tried the brackets but was missing the tag and value parms.
